File: platform/util/src/com/intellij/openapi/util/SafeStAXStreamBuilder.java
// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.openapi.util;

import org.codehaus.stax2.XMLStreamReader2;
import org.jdom.Document;
import org.jdom.Element;
import org.jdom.Namespace;
import org.jdom.Verifier;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;

import javax.xml.stream.XMLStreamException;

import static javax.xml.stream.XMLStreamConstants.*;

// DTD, COMMENT and PROCESSING_INSTRUCTION are ignored
@ApiStatus.Internal
public final class SafeStAXStreamBuilder {
  public static final SafeJdomFactory FACTORY = new SafeJdomFactory.BaseSafeJdomFactory();

  static Document buildDocument(@NotNull XMLStreamReader2 stream) throws XMLStreamException {
    int state = stream.getEventType();

    if (START_DOCUMENT != state) {
      throw new XMLStreamException("JDOM requires that XMLStreamReaders are at their beginning when being processed.");
    }

    Document document = new Document();

    while (state != END_DOCUMENT) {
      switch (state) {
        case START_DOCUMENT:
          // for the <?xml version="..." standalone=".."?>
          document.setBaseURI(stream.getLocation().getSystemId());
          document.setProperty("ENCODING_SCHEME", stream.getCharacterEncodingScheme());
          document.setProperty("STANDALONE", String.valueOf(stream.isStandalone()));
          document.setProperty("ENCODING", stream.getEncoding());
          break;

        case DTD:
        case COMMENT:
        case PROCESSING_INSTRUCTION:
        case SPACE:
          break;

        case START_ELEMENT:
          document.setRootElement(processElementFragment(stream, true, true, FACTORY));
          break;

        case CHARACTERS:
          String badTxt = stream.getText();
          if (!Verifier.isAllXMLWhitespace(badTxt)) {
            throw new XMLStreamException("Unexpected XMLStream event at Document level: CHARACTERS (" + badTxt + ")");
          }
          // otherwise, ignore the chars
          break;

        default:
          throw new XMLStreamException("Unexpected XMLStream event at Document level:" + state);
      }

      if (stream.hasNext()) {
        state = stream.next();
      }
      else {
        throw new XMLStreamException("Unexpected end-of-XMLStreamReader");
      }
    }
    return document;
  }

  public static Element build(@NotNull XMLStreamReader2 stream,
                              @SuppressWarnings("SameParameterValue") boolean isIgnoreBoundaryWhitespace,
                              boolean isNsSupported,
                              @NotNull SafeJdomFactory factory) throws XMLStreamException {
    int state = stream.getEventType();

    if (state != START_DOCUMENT) {
      throw new XMLStreamException("JDOM requires that XMLStreamReaders are at their beginning when being processed");
    }

    Element rootElement = null;
    while (state != END_DOCUMENT) {
      switch (state) {
        case START_DOCUMENT:
        case SPACE:
        case CHARACTERS:
        case COMMENT:
        case PROCESSING_INSTRUCTION:
        case DTD:
          break;

        case START_ELEMENT:
          rootElement = processElementFragment(stream, isIgnoreBoundaryWhitespace, isNsSupported, factory);
          break;

        default:
          throw new XMLStreamException("Unexpected XMLStream event " + state);
      }

      if (stream.hasNext()) {
        state = stream.next();
      }
      else {
        throw new XMLStreamException("Unexpected end-of-XMLStreamReader");
      }
    }

    if (rootElement == null) {
      // to avoid NPE
      return new Element("empty");
    }
    return rootElement;
  }

  public static @NotNull Element processElementFragment(@NotNull XMLStreamReader2 reader,
                                                        boolean isIgnoreBoundaryWhitespace,
                                                        boolean isNsSupported,
                                                        @NotNull SafeJdomFactory factory) throws XMLStreamException {
    Element fragment = processElement(reader, isNsSupported, factory);
    Element current = fragment;
    int depth = 1;
    while (depth > 0 && reader.hasNext()) {
      switch (reader.next()) {
        case START_ELEMENT:
          Element tmp = processElement(reader, isNsSupported, factory);
          current.addContent(tmp);
          current = tmp;
          depth++;
          break;
        case END_ELEMENT:
          current = current.getParentElement();
          depth--;
          break;
        case CDATA:
          current.addContent(factory.cdata(reader.getText()));
          break;

        case SPACE:
          if (!isIgnoreBoundaryWhitespace) {
            current.addContent(factory.text(reader.getText()));
          }
          break;

        case CHARACTERS:
          if (!isIgnoreBoundaryWhitespace || !reader.isWhiteSpace()) {
            current.addContent(factory.text(reader.getText()));
          }
          break;

        case ENTITY_REFERENCE:
        case COMMENT:
        case PROCESSING_INSTRUCTION:
          break;

        default:
          throw new XMLStreamException("Unexpected XMLStream event " + reader.getEventType(), reader.getLocation());
      }
    }

    return fragment;
  }

  private static @NotNull Element processElement(@NotNull XMLStreamReader2 reader,
                                                 boolean isNsSupported,
                                                 @NotNull SafeJdomFactory factory) {
    Element element = factory.element(reader.getLocalName(), isNsSupported
                                                             ? Namespace.getNamespace(reader.getPrefix(), reader.getNamespaceURI())
                                                             : Namespace.NO_NAMESPACE);
    // handle attributes
    for (int i = 0, len = reader.getAttributeCount(); i < len; i++) {
      element.setAttribute(factory.attribute(
        reader.getAttributeLocalName(i),
        reader.getAttributeValue(i),
        isNsSupported ? Namespace.getNamespace(reader.getAttributePrefix(i), reader.getAttributeNamespace(i)) : Namespace.NO_NAMESPACE
      ));
    }

    if (isNsSupported) {
      // handle namespaces
      for (int i = 0, len = reader.getNamespaceCount(); i < len; i++) {
        element.addNamespaceDeclaration(Namespace.getNamespace(reader.getNamespacePrefix(i), reader.getNamespaceURI(i)));
      }
    }

    return element;
  }
}

File: platform/util/src/com/intellij/openapi/util/SafeStAXStreamBuilder.kt
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
@file:ApiStatus.Internal
package com.intellij.openapi.util

import com.intellij.util.xml.dom.createXmlStreamReader
import org.codehaus.stax2.XMLStreamReader2
import org.jdom.*
import org.jetbrains.annotations.ApiStatus
import java.io.Reader
import java.nio.file.Files
import java.nio.file.Path
import javax.xml.stream.XMLStreamConstants
import javax.xml.stream.XMLStreamException

// DTD, COMMENT and PROCESSING_INSTRUCTION are ignored

@Throws(XMLStreamException::class)
@ApiStatus.Internal
fun buildJdomDocument(stream: XMLStreamReader2): Document {
  var state = stream.eventType

  if (XMLStreamConstants.START_DOCUMENT != state) {
    throw XMLStreamException("JDOM requires that XMLStreamReaders are at their beginning when being processed.")
  }

  val document = Document()

  while (state != XMLStreamConstants.END_DOCUMENT) {
    when (state) {
      XMLStreamConstants.START_DOCUMENT -> {
        // for the <?xml version="..." standalone=".."?>
        document.baseURI = stream.location.systemId
        document.setProperty("ENCODING_SCHEME", stream.characterEncodingScheme)
        document.setProperty("STANDALONE", stream.isStandalone.toString())
        document.setProperty("ENCODING", stream.encoding)
      }
      XMLStreamConstants.DTD, XMLStreamConstants.COMMENT, XMLStreamConstants.PROCESSING_INSTRUCTION, XMLStreamConstants.SPACE -> {}
      XMLStreamConstants.START_ELEMENT -> {
        document.setRootElement(processElementFragment(reader = stream, isNsSupported = true))
      }
      XMLStreamConstants.CHARACTERS -> {
        val badTxt = stream.text
        if (!Verifier.isAllXMLWhitespace(badTxt)) {
          throw XMLStreamException("Unexpected XMLStream event at Document level: CHARACTERS ($badTxt)")
        }
      }
      else -> throw XMLStreamException("Unexpected XMLStream event at Document level: $state")
    }

    if (stream.hasNext()) {
      state = stream.next()
    }
    else {
      throw XMLStreamException("Unexpected end-of-XMLStreamReader")
    }
  }
  return document
}

@Throws(XMLStreamException::class)
@ApiStatus.Internal
fun buildNsUnawareJdom(reader: Reader): Element {
  val xmlStreamReader = createXmlStreamReader(reader)
  try {
    return buildNsUnawareJdomAndClose(xmlStreamReader)
  }
  finally {
    xmlStreamReader.close()
  }
}

@Throws(XMLStreamException::class)
@ApiStatus.Internal
fun buildNsUnawareJdom(data: ByteArray): Element {
  val xmlStreamReader = createXmlStreamReader(data)
  try {
    return buildNsUnawareJdomAndClose(xmlStreamReader)
  }
  finally {
    xmlStreamReader.close()
  }
}

@Throws(XMLStreamException::class)
@ApiStatus.Internal
fun buildNsUnawareJdom(file: Path): Element {
  val xmlStreamReader = createXmlStreamReader(Files.newInputStream(file))
  try {
    return buildNsUnawareJdomAndClose(xmlStreamReader)
  }
  finally {
    xmlStreamReader.close()
  }
}

@Throws(XMLStreamException::class)
@ApiStatus.Internal
fun buildNsUnawareJdomAndClose(stream: XMLStreamReader2): Element {
  return buildJdom(stream = stream, isNsSupported = false)
}

@Throws(XMLStreamException::class)
@ApiStatus.Internal
fun buildJdom(
  stream: XMLStreamReader2,
  isNsSupported: Boolean,
): Element {
  var state = stream.eventType
  if (state != XMLStreamConstants.START_DOCUMENT) {
    throw XMLStreamException("JDOM requires that XMLStreamReaders are at their beginning when being processed")
  }

  var rootElement: Element? = null
  while (state != XMLStreamConstants.END_DOCUMENT) {
    when (state) {
      XMLStreamConstants.START_DOCUMENT,
      XMLStreamConstants.SPACE,
      XMLStreamConstants.CHARACTERS,
      XMLStreamConstants.COMMENT,
      XMLStreamConstants.PROCESSING_INSTRUCTION,
      XMLStreamConstants.DTD -> {
      }
      XMLStreamConstants.START_ELEMENT -> {
        rootElement = processElementFragment(reader = stream, isNsSupported = isNsSupported)
      }
      else -> throw XMLStreamException("Unexpected XMLStream event $state")
    }

    if (stream.hasNext()) {
      state = stream.next()
    }
    else {
      throw XMLStreamException("Unexpected end-of-XMLStreamReader")
    }
  }

  if (rootElement == null) {
    // to avoid NPE
    return Element("empty")
  }
  return rootElement
}

@Throws(XMLStreamException::class)
private fun processElementFragment(
  reader: XMLStreamReader2,
  isNsSupported: Boolean,
): Element {
  val fragment = processElement(reader = reader, isNsSupported = isNsSupported)
  var current = fragment
  var depth = 1
  while (depth > 0 && reader.hasNext()) {
    when (reader.next()) {
      XMLStreamConstants.START_ELEMENT -> {
        val tmp = processElement(reader = reader, isNsSupported = isNsSupported)
        current.addContent(tmp)
        current = tmp
        depth++
      }
      XMLStreamConstants.END_ELEMENT -> {
        current = current.parentElement ?: return fragment
        depth--
      }
      XMLStreamConstants.CDATA -> current.addContent(CDATA(reader.text))
      XMLStreamConstants.SPACE -> {
      }
      XMLStreamConstants.CHARACTERS -> {
        if (!reader.isWhiteSpace) {
          current.addContent(Text(reader.text))
        }
      }
      XMLStreamConstants.ENTITY_REFERENCE, XMLStreamConstants.COMMENT, XMLStreamConstants.PROCESSING_INSTRUCTION -> {}
      else -> throw XMLStreamException("Unexpected XMLStream event ${reader.eventType}", reader.location)
    }
  }

  return fragment
}

private fun processElement(reader: XMLStreamReader2, isNsSupported: Boolean): Element {
  val element = Element(reader.localName, if (isNsSupported) {
    Namespace.getNamespace(reader.prefix, reader.namespaceURI)
  }
  else {
    Namespace.NO_NAMESPACE
  })

  // handle attributes
  val attributeCount = reader.attributeCount
  if (attributeCount != 0) {
    val list = element.initAttributeList(attributeCount)
    for (i in 0 until attributeCount) {
      list.doAdd(Attribute(
        true,
        reader.getAttributeLocalName(i),
        reader.getAttributeValue(i),
        if (isNsSupported) Namespace.getNamespace(reader.getAttributePrefix(i), reader.getAttributeNamespace(i)) else Namespace.NO_NAMESPACE,
      ))
    }
  }

  if (isNsSupported) {
    // handle namespaces
    var i = 0
    val len = reader.namespaceCount
    while (i < len) {
      element.addNamespaceDeclaration(Namespace.getNamespace(reader.getNamespacePrefix(i), reader.getNamespaceURI(i)))
      i++
    }
  }

  return element
}